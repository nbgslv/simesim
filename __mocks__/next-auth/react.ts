const mockSession = {
  expires: new Date(Date.now() + 2 * 86400).toISOString(),
  user: {
    email: '0512345678',
    emailEmail: 'test@test.com',
    id: 'clcclcclcclcclcclc1234',
    image: null,
    name: 'foo bar',
    role: 'ADMIN',
  },
};

export const useSession = jest.fn(
  () => ({ data: mockSession, status: 'authenticated' }) // return type is [] in v3 but changed to {} in v4
);
export const signOut = jest.fn(() =>
  Promise.resolve({ url: process.env.NEXTAUTH_URL as string })
);
