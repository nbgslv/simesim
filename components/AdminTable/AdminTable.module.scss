.main {
  height: 83%;
}

.table {
  height: 100%;
}
