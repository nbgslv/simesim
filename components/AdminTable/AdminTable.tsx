import React, { RefObject, useEffect, useMemo, useRef } from 'react';
import {
  DataGrid,
  GridColumns,
  GridEventListener,
  GridRowId,
  GridRowModel,
  GridRowModesModel,
  GridRowParams,
  GridSelectionModel,
  GridValidRowModel,
  MuiEvent,
} from '@mui/x-data-grid';
import { GridApiCommunity } from '@mui/x-data-grid/internals';
import styles from './AdminTable.module.scss';
import EditToolbar from './EditToolbar';
import useAddActions from './useAddActions';

type AdminTableProps<T extends GridValidRowModel> = {
  data: GridRowModel<T>[];
  title?: string;
  handleRowsSelection?: (ids: GridSelectionModel) => boolean;
  processRowUpdate?: (newRow: T, oldRow: T) => Promise<any>;
  columns: GridColumns;
  multiActions?: string[];
  rowActions?: string[];
  limit?: number;
  editMode?: 'row' | 'cell';
  addRow?: () => Promise<
    { id: GridRowId; columnToFocus: string | undefined } | undefined | Error
  >;
  onDelete?: (ids: GridSelectionModel) => Promise<{ ids: GridSelectionModel }>;
  deleteRows?: (ids: GridRowId[]) => Promise<void>;
};

function useApiRef({ columns }: { columns: any }) {
  const apiRef = useRef<GridApiCommunity>(null);
  // eslint-disable-next-line @typescript-eslint/naming-convention,no-underscore-dangle
  const _columns = useMemo(
    () =>
      columns.concat({
        field: '__HIDDEN__',
        width: 0,
        hide: true,
        renderCell: (params: any) => {
          // @ts-ignore
          apiRef.current = params.api;
          return null;
        },
      }),
    [columns]
  );

  return { apiRef, columns: _columns };
}

const AdminTable = <T extends GridValidRowModel>({
  data,
  title,
  handleRowsSelection,
  processRowUpdate,
  columns,
  multiActions = ['delete', 'add'],
  rowActions = ['delete', 'edit'],
  limit = 15,
  editMode = 'cell',
  addRow,
  deleteRows,
}: AdminTableProps<T>) => {
  const [dataRows, setDataRows] = React.useState<GridRowModel<T>[]>(data);
  const [
    selectionModel,
    setSelectionModel,
  ] = React.useState<GridSelectionModel>([]);
  const [rowModesModel, setRowModesModel] = React.useState<GridRowModesModel>(
    {}
  );
  const {
    columns: columnsWithApi,
  }: { apiRef: RefObject<GridApiCommunity>; columns: GridColumns } = useApiRef({
    columns,
  });
  const {
    columns: columnsWithActions,
  }: { columns: GridColumns } = useAddActions({
    columns: columnsWithApi,
    rowsActions: rowActions,
    rowModesModel,
    setRowModesModel,
    setRows: setDataRows,
    rows: dataRows,
  });

  useEffect(() => {
    setDataRows(data);
  }, [data]);

  useEffect(() => {
    if (
      Array.isArray(selectionModel) &&
      selectionModel.length > 0 &&
      handleRowsSelection
    ) {
      const deselectRows = handleRowsSelection(selectionModel);
      if (deselectRows) {
        setSelectionModel([]);
      }
    }
  }, [selectionModel]);

  const handleRowEditStart = (
    params: GridRowParams,
    event: MuiEvent<React.SyntheticEvent>
  ) => {
    // eslint-disable-next-line no-param-reassign
    event.defaultMuiPrevented = true;
  };

  const handleRowEditStop: GridEventListener<'rowEditStop'> = (
    params,
    event
  ) => {
    // eslint-disable-next-line no-param-reassign
    event.defaultMuiPrevented = true;
  };

  return (
    <div className={styles.main}>
      {title && <h2>{title}</h2>}
      <div className={styles.table} dir="ltr">
        <DataGrid
          pageSize={limit}
          pagination
          columns={rowActions.length ? columnsWithActions : columnsWithApi}
          rows={dataRows}
          processRowUpdate={processRowUpdate}
          checkboxSelection
          disableSelectionOnClick
          selectionModel={selectionModel}
          onSelectionModelChange={(newSelectionModel) =>
            setSelectionModel(newSelectionModel)
          }
          editMode={editMode}
          rowModesModel={rowModesModel}
          onRowModesModelChange={(newModel) => setRowModesModel(newModel)}
          onRowEditStart={handleRowEditStart}
          onRowEditStop={handleRowEditStop}
          components={{
            Toolbar: multiActions.length ? EditToolbar : null,
          }}
          componentsProps={{
            toolbar: {
              selectedRows: selectionModel,
              setRows: setDataRows,
              setRowModesModel,
              multiActions,
              addRow,
              deleteRows,
            },
          }}
          experimentalFeatures={{ newEditingApi: true }}
        />
      </div>
    </div>
  );
};

export default AdminTable;
