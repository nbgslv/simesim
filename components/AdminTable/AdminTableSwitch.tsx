import React from 'react';
import { Spinner, Form } from 'react-bootstrap';
import { GridRowId, GridRowModel, GridValidRowModel } from '@mui/x-data-grid';

type AdminTableSwitchProps<T extends GridValidRowModel> = {
  checked: boolean;
  onChange: (checked: boolean, rowId: GridRowId, row: GridRowModel<T>) => void;
  rowId: GridRowId;
  row: GridRowModel<T>;
  loading: GridRowId;
};

const AdminTableSwitch = <T extends GridValidRowModel>({
  checked,
  onChange,
  rowId,
  row,
  loading,
}: AdminTableSwitchProps<T>) => {
  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    onChange(event.target.checked, rowId, row);
  };

  if (loading === rowId)
    return <Spinner className="mx-auto" animation={'border'} />;
  return <Form.Check type="switch" checked={checked} onChange={handleChange} />;
};

export default AdminTableSwitch;
