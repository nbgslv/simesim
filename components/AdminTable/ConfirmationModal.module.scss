@import 'styles/variables';

.main {
  :global * {
    color: $text-color;
  }
}
