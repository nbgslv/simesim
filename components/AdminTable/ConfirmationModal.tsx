import React from 'react';
import { Modal, Button } from 'react-bootstrap';
import styles from './ConfirmationModal.module.scss';

export type ConfirmationModalProps = {
  show: boolean;
  onHide?: () => void;
  title: string;
  body: string;
  confirmButtonText?: string;
  cancelButtonText?: string;
  confirmAction?: (() => void) | (() => Promise<void>);
  cancelAction?: (() => void) | (() => Promise<void>);
  id?: string;
};

const ConfirmationModal = ({
  show,
  // onHide,
  title,
  body,
  confirmButtonText,
  cancelButtonText,
  confirmAction,
  cancelAction,
  id,
  ...props
}: ConfirmationModalProps) => (
  <Modal
    id={id}
    show={show}
    onHide={cancelAction}
    className={styles.main}
    {...props}
  >
    <Modal.Header closeButton>
      <Modal.Title>{title}</Modal.Title>
    </Modal.Header>

    <Modal.Body>
      <p>{body}</p>
    </Modal.Body>

    <Modal.Footer>
      {cancelButtonText && (
        <Button onClick={cancelAction} variant="secondary">
          {cancelButtonText}
        </Button>
      )}
      {confirmButtonText && (
        <Button onClick={confirmAction} variant="primary">
          {confirmButtonText}
        </Button>
      )}
    </Modal.Footer>
  </Modal>
);

export default ConfirmationModal;
