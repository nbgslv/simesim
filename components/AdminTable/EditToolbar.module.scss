@import 'styles/variables';

.button {
  margin-left: 1rem;

  :global svg {
    margin-left: 1rem;
  }

  :global path {
    color: $background-color;
  }
}

.spinner {
  position: absolute;
  right: 50%;
  transform: translateX(-50%);
}
