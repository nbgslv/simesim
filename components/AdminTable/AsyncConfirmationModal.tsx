import NiceModal, { useModal } from '@ebay/nice-modal-react';
import ConfirmationModal, { ConfirmationModalProps } from './ConfirmationModal';

export default NiceModal.create(
  ({
    show, // onHide,
    title,
    body,
    confirmButtonText,
    cancelButtonText,
    confirmAction,
    cancelAction,
    id,
    ...props
  }: ConfirmationModalProps) => {
    const modal = useModal();
    return (
      <ConfirmationModal
        show={show}
        title={title}
        body={body}
        confirmButtonText={confirmButtonText}
        cancelButtonText={cancelButtonText}
        confirmAction={() => {
          modal.resolve(true);
          confirmAction?.();
        }}
        cancelAction={() => {
          modal.reject(false);
          cancelAction?.();
        }}
        id={id}
        {...props}
      />
    );
  }
);
