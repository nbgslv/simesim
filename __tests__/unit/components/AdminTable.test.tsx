import React from 'react';
import { render } from '@testing-library/react';
import { DataGrid, GridColumns, GridRowModel } from '@mui/x-data-grid';
import AdminTable from '../../../components/AdminTable/AdminTable';

const handleRowsSelection = jest.fn((ids) => ids);

jest.mock('@mui/x-data-grid/DataGrid', () => ({
  ...jest.requireActual('@mui/x-data-grid/DataGrid'),
  DataGrid: jest.fn(
    ({
      rows,
      columns,
    }: {
      rows: GridRowModel[];
      columns: GridColumns<any>;
    }) => (
      <table>
        <thead>
          <tr>
            <th />
            {columns.map((column) => (
              <th key={column.field}>{column.headerName}</th>
            ))}
          </tr>
        </thead>
        <tbody>
          {rows.map((row) => (
            <tr key={row.id}>
              <td>
                <input
                  type="checkbox"
                  onClick={() => jest.fn(() => handleRowsSelection([row.id]))}
                />
              </td>
              {columns.map((column) => (
                <td key={column.field}>{row[column.field]}</td>
              ))}
            </tr>
          ))}
        </tbody>
      </table>
    )
  ),
}));

const mockedDataGrid = jest.mocked(DataGrid);

describe('AdminTable', () => {
  const columns = [
    { field: 'id', headerName: 'ID', width: 70 },
    { field: 'name', headerName: 'Name', width: 130 },
    { field: 'age', headerName: 'Age', type: 'number', width: 90 },
  ];
  const rows = [
    { id: 1, name: 'John Smith', age: 35 },
    { id: 2, name: 'Jane Doe', age: 25 },
    { id: 3, name: 'Bob Johnson', age: 45 },
  ];

  it('matches snapshot', () => {
    const { container } = render(
      <AdminTable
        disableVirtualization
        rowActions={[]}
        data={rows}
        columns={columns}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders basic table without errors', () => {
    render(
      <AdminTable
        disableVirtualization
        rowActions={[]}
        data={rows}
        columns={columns}
      />
    );
  });

  it('displays table rows and columns', () => {
    render(<AdminTable disableVirtualization data={rows} columns={columns} />);
    expect(mockedDataGrid).toHaveBeenLastCalledWith(
      expect.objectContaining({
        rows: expect.arrayContaining(rows),
        columns: expect.arrayContaining(columns),
      }),
      {}
    );
  });

  it('should display title', () => {
    const { getByText } = render(
      <AdminTable
        disableVirtualization
        title="Test Title"
        data={rows}
        columns={columns}
      />
    );
    expect(getByText('Test Title')).toBeInTheDocument();
  });
});
