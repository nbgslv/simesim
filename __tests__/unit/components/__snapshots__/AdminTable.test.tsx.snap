// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminTable matches snapshot 1`] = `
<div>
  <div
    class="main"
  >
    <div
      class="table"
      dir="ltr"
    >
      <table>
        <thead>
          <tr>
            <th />
            <th>
              ID
            </th>
            <th>
              Name
            </th>
            <th>
              Age
            </th>
            <th />
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <input
                type="checkbox"
              />
            </td>
            <td>
              1
            </td>
            <td>
              John Smith
            </td>
            <td>
              35
            </td>
            <td />
          </tr>
          <tr>
            <td>
              <input
                type="checkbox"
              />
            </td>
            <td>
              2
            </td>
            <td>
              Jane Doe
            </td>
            <td>
              25
            </td>
            <td />
          </tr>
          <tr>
            <td>
              <input
                type="checkbox"
              />
            </td>
            <td>
              3
            </td>
            <td>
              Bob Johnson
            </td>
            <td>
              45
            </td>
            <td />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
